### DISCUSSION TITLE
Roadmap for the next year? (year 3)

### DISCUSSION THREAD

**Discussion by leossmith** [airsdk/Adobe-Runtime-Support#1472] [General]
So another year is about to end and it is time to check what 2022 will bring us. Following my previous year question #597, does Harman have a roadmap to share with us for the next year and what we should expect from AIR? Maybe it would be a great idea to have a list of the roadmap (example of [gm roadmap](https://www.yoyogames.com/en/roadmap#/extensions)).

Are there any plans of promoting AIR? Adding new features?

**Comment by dimensionscape** [comment]
Maybe something with a path that doesn't end with EOL perhaps?

**Comment by JoNMii** [comment]
Ded

**Comment by ajwfrost** [comment]
Hi

Well we're still working on some of the plans from last year - multimedia updates being the main one here that has been delayed. We have the new framework running media files but need to patch them into the display output properly..

The other main updates planned are:
* Simplification of 'getting started' with management of the SDK and updates; restructuring the SDK and where/how you change the configuration settings
* Debugging improvements including "edit & continue" or hot-reload for on-device debugging
* AS3 API improvements including better screen information and control (cut-out info is still WIP, plus z-ordering of stagex elements)
* Restructure the runtime loop to improve things like "waiting for GPU", Android ANR reports, etc
* Android AOT compilation
* Integration with third party components (i.e. extending ANE interfaces including a way to embed external apps into the display list .. this is very much at a 'proof of concept' level to start with but we have it working with OpenGL ES via EGLImage...)

Please keep the suggestions coming in..

thanks

**Reply by leossmith** [reply]
Android AOT sounds exciting and it might actually solve a lot of issues on android. A hot-reload would be also more than welcome as it makes the development time a lot faster.

**Reply by Elintondm** [reply]
Anything to reduce the ANR rate would be great!!!

**Reply by toysonlocation** [reply]
Regarding media playback, what I'd really really like to see, because I can do it natively, is to media.seek(t) correctly. Right now I get very different behavior between iOS (in a word: works) and Android (in a contracted word: doesn't), even when I export my h264 with all i-frames (keyframes). I sent you guys a test video a ways back that is frame-by-frame. In fact the behavior between the platforms is so different I have to use an ANE for Android, and Krazy Kode on iOS. Let me know if you need the video sample again.

**Reply by 2jfw** [reply]
Any chance for getting an "official" Roadmap (section) on "https://airsdk.harman.com/" (also https://airsdk.dev/) ?
Actively maintained Roadmaps are always welcome and a healthy sign, casting a positive light on the project :)

**Reply by marchbold** [reply]
@2jfw Do you have any examples of other framework roadmaps that might be worth referencing? Just interested to see how other platforms handle this.

**Reply by 2jfw** [reply]
@marchbold: Most of the Roadmaps I see are actually image-wise Road Maps from games on Steam so I dont have any example at hand I could share spontaneously.

I did some quick search for how JetBrains do it and saw something like this:
https://www.jetbrains.com/de-de/space/roadmap/

I could also imagine something like in "Trello" style:
https://trello.com/b/vtRJoZQp/dayz-frostbite-roadmap

Or Unity's Roadmap:
https://unity.com/de/roadmap/unity-platform

**Reply by leossmith** [reply]
@marchbold 

[This ](https://www.yoyogames.com/en/roadmap#/features)is also a nice example of a roadmap.

**Reply by marchbold** [reply]
Thanks all, will have a look. I suppose if we can put together a framework for Harman on the airsdk site that they just have to populate.

Anyone seen any good ones using github "Projects" ? Maybe we could use that seeing as there is a lot of interaction around this repository now?

**Reply by 2jfw** [reply]
Just found this one here (I'm calling the look "Trello" Style, since I first ever saw it there like this):
https://github.com/github/roadmap/projects/1 (https://github.com/orgs/github/projects/4247)

Not sure if we could utilize it somehow?
Also not sure if Harman would be willing to actively work with a new  github project using _Issues_ as the **tasks** for the road map?

**Reply by marchbold** [reply]
Yeah I figure the best way to get a good road map is to make it as simple as possible to maintain, and I figure keeping everything in one place will make it easier? It's definitely "trello" style and something we can turn on in this repository if needed.

**Reply by ajwfrost** [reply]
@2jfw that looks good ... and yes @marchbold it seem that uses the 'Projects' mechanism.

I was going to say that whatever we did, it would be simplest if we could have a way to link things back here to issues (or discussions) since that's often where the detail will be. So if we could get that sort of thing set up whereby we then add some sort of tags to an issue, so it would appear in this roadmap list? Not entirely sure how it works, we don't want the whole set of open issues to be appearing in a roadmap document!!

thanks

**Reply by marchbold** [reply]
@ajwfrost  Well if linking to the issues here is a priority then GitHub Projects definitely are an easy thing to start with. I'll get this setup over the next couple of weeks. We'll probably add a few more labels and things to make it all work well. Issues definitely aren't automatically added to it, you create a "card" from an issue and then can arrange them in "columns" .  

I did start a basic one for APM as a simple reference:

https://github.com/airsdk/apm/projects/1

**Reply by 2jfw** [reply]
@marchbold Looks good :) In the second link from above I saw that you are also able to _change_ the layout from **Board** to **Table** (-> see the arrow on the first tab "Board" and change to "Table") - or have a look here: https://github.com/orgs/github/projects/4247/views/1?layout=table
Any chance to also have this _useful_ layout available for the Roadmap?

![image](https://user-images.githubusercontent.com/73781224/149473317-d7b7e422-782b-49bd-b44e-a29ae202ea2b.png)

**Reply by marchbold** [reply]
@2jfw Yeah I think that's the new "Projects beta" interface that is in development? Seems to be enabled for us so we definitely could use that too.

**Reply by 2jfw** [reply]
@marchbold Yes, you are right! I was not quite able to put the "Beta" in any context until now ..

**Reply by marchbold** [reply]
@ajwfrost I've setup this : https://github.com/orgs/airsdk/projects/2

Let me know if it's worth keeping up there? Looks like it could be quite a useful tool to display progress / planned development

**Reply by ajwfrost** [reply]
Thanks @marchbold  - the one thing I hadn't quite worked out was how we would create/assign issues to different quarters, but since you'd added one item already we could work it out (at least, a way to edit an existing issue to assign it to a 'project' and then give it a 'status'...)

Looks good :-)

**Reply by marchbold** [reply]
Just sort of copied the GitHub one there which seemed like a decent way to do it?

**Reply by 2jfw** [reply]
Not 100% sure but I would have thought that the **Milestone** property would be the correct approach/attribute for sorting/assigning the issue to a "quarter"

**Reply by marchbold** [reply]
That definitely is the traditional approach with github, we use them a bit for our extensions. But milestones are a bit more rigid than the projects for arranging etc. Have a look at the github one you linked earlier and you'll see they aren't using them either. Also milestones are per repo whereas these boards and the status flags can cross repos so eg they could have improvements in the other repos included in the roadmap (apm/ANEs etc) 

But doesn't bother me if Harman do want to use them.

**Reply by 2jfw** [reply]
Alright - thank you for the clarification, Michael!

**Comment by M4rkY4** [comment]
Hi, @ajwfrost 

> Please keep the suggestions coming in..

few ideas for the future of AIR:

**1)** *.has (Harman Action Script) file format
a file with code with AS3-like syntax, but with added syntax sugar:

arrow functions:
set=> get=>

listener +=
listener -=

object?.method(); // shorter than "if (object) object.method();"

Before compile, it generates *.as file with converted code to current AS3 or is an alternative to *.as file.
Thus you will avoid a need to create new virtual machine for AS4.
I hope, Intellij IDEA flash plugin will be updated to use that syntax.


**2)** Better/easier Multithreading


**3)** Windows terminal max characters length fix (use compile parameters from a text file, not from single terminal command)


**4)** Ask Adobe to add texture atlas folders to AnimateCC.

**_Problem:_**
Starling is great, but manual objects positioning, exporting/importing animations, manual objects disposing and etc. are too outdated approaches. And I see no advantages of Starling comparing to Defold, Phaser.io or even Unity (Unity code syntax can be adapted to be more 2d-friendly, a bit of work, but it greatly pays off).
GPU rendering, on the other hand, was amazing. When you create games inside Animate, it's by far the fastest and most convenient way to create a game! And GPU rendering performance was great too.
The problem with GPU rendering, as far as I know, is that if you create too many graphics for a single atlas (which seems to be auto-generated), the game auto-generates multiple atlases the way it wants, so FPS drops significantly as of too much uncontrolled draw call. At least that's what I've read approx. 5 years ago.

**_What I propose to ask Adobe:_**
In Animate CC library, you can create a folder of type "Atlas folder" and all bitmaps inside that folder should be packed to a separate auto-generated texture atlas. If that separate UI icon is too complicated for Adobe, ask them not to add special icon for that folder, but if folder name starts from "ATLAS_" and is on a root of library, then internally pack bitmaps of that folder to a separate atlas.
In that way we can control draw calls and that would be amazing, flash DisplayList will become the best Gamedev tool again, I hope not only for me. And it gives a reason to start new projects with flash/air technology.

If it would be possible to regenerate atlas after swf was created by AnimateCC, that would be great alternative without asking Adobe.

**Reply by ajwfrost** [reply]
Just on the first of these:

> *.has (Harman Action Script) file format
> a file with code with AS3-like syntax, but with added syntax sugar
> Before compile, it generates *.as file with converted code to current AS3 or is an alternative to *.as file.
> Thus you will avoid a need to create new virtual machine for AS4

Not sure whether this would completely work: the idea here is that we have new ActionScript syntaxes but that it then generates 'normal' AS3 code. But that would probably be more effort than just updating the AS3 compiler..! And if it's possible for these to run within the existing virtual machine then we hit limitations immediately. For example the use of a 64-bit integer type .. it's possible to emulate it, of course, but there seems little value in that; but to actually support it, it would need to be supported within the virtual machine as well (and the ABC format).

So I'm not saying that we won't do things, just that it's more likely that we will gradually evolve the existing language and vm together. Hopefully a lot of things could be solved just from the compiler, but there will also be VM updates at which point a file that's compiled with a newer compiler will need to be run on the newer VM...

> arrow functions:
> set=> get=>

I'm not 100% sure what you mean; whether this is a general "arrow function" syntax being available for anonymous functions (e.g. `addListener("timer", (e)=>{ trace(e); doSomething(); });`) - or whether you mean specifically for a property set/get pair, such as `public property name : String { set=>_name = value; get=>return _name; } private var _name : String;` which would be a bit like C#?

> listener +=
> listener -=

An example might help here too .. e.g. `stage.addEventListener(Mouse.CLICK, myClickHandler);` would become ..? `stage.listeners[Mouse.CLICK] += myClickHandler;`? This might be a possibility (at least if we had operator overloading!)

> object?.method(); // shorter than "if (object) object.method();"

This one is already planned and should be fairly easy to do; at least as long as we can work out what happens if it's used in an expression (i.e. `object?.property = val;` and `val = object?.property;` - the latter should set `val` to `null` if `object` is null, the former perhaps would equate to `if (object) object.property = val;` but is a bit confusing perhaps...


It will be interesting to see how far we can go with the syntax updates - and with AVM+ and ABC changes....! Quite an interesting area :-)

thanks

**Reply by M4rkY4** [reply]
Wow, that's great that some of these are already planned!

Certainly, you have a better vision on how to implement those features, I'm just missing them and though my ideas might be easier to implement, which seem not to be the case :)

Concerning instances, I meant the second one, or something like:
```
  public static get instance():Banana => _instance;
  public static set instance():Banana => _instance = value;
```
maybe even omitting "function" word
or
`  public static function currentBanana():Banana { get => _currentBanana || new Banana(); set; }`
but any/both variants you've provided would be great!

Concerning events, my bad, I've used uncommon and manually created events and got used to them, with current flash events it makes no sense :)

Though, the most viable feature for me is to make GPU rendering more useful in big projects with somehow controlling draw calls.

Thanks, your plans look very inspiring to me!
